### PR TITLE
Refactor .yml and .yml.erb handling

### LINKED ISSUE #26: ERb YAML templates should have the .erb extension
- Without the extension, code editors will detect it as YAML and will not highlight ERb code blocks
- The file may not be valid YAML with the ERb code intact, making something like `YAML.load(File.read('xyz.yml'))` fail.
- Having the extension makes it simple to conditionally interpolate only when necessary, the glob should therefore be `*.{yml,yml.erb}` and the reader should only run ERb interpolation when the extension is present.

### LINKED ISSUE #199: Pharos::Kube#apply_stack does not sort resource files
Testing with `docker-compose run pharos-cluster up` with ruby 2.4.3, the results of `Dir.glob` are not sorted. This causes stack resources to fail when applied in the wrong order:

```
==> addons: 167.99.36.141
    Applying addon ingress-nginx ...
kube stack ingress-nginx files from glob: ["/app/lib/pharos/resources/ingress-nginx/default-backend-service.yml", "/app/lib/pharos/resources/ingress-nginx/09-udp-services-configmap.yml", "/app/lib/pharos/resources/ingress-nginx/08-tcp-services-configmap.yml", "/app/lib/pharos/resources/ingress-nginx/06-role-binding.yml", "/app/lib/pharos/resources/ingress-nginx/05-cluster-role-binding.yml", "/app/lib/pharos/resources/ingress-nginx/04-service-account.yml", "/app/lib/pharos/resources/ingress-nginx/03-role.yml", "/app/lib/pharos/resources/ingress-nginx/02-cluster-role.yml", "/app/lib/pharos/resources/ingress-nginx/01-namespace.yml", "/app/lib/pharos/resources/ingress-nginx/07-configmap.yml", "/app/lib/pharos/resources/ingress-nginx/daemonset.yml", "/app/lib/pharos/resources/ingress-nginx/default-backend.yml"]
Module: apply resource: v1/Service/default-http-backend
namespaces "ingress-nginx" not found
/usr/local/bundle/gems/kubeclient-3.0.0/lib/kubeclient/common.rb:116:in `rescue in handle_exception'
/usr/local/bundle/gems/kubeclient-3.0.0/lib/kubeclient/common.rb:106:in `handle_exception'
/usr/local/bundle/gems/kubeclient-3.0.0/lib/kubeclient/common.rb:325:in `create_entity'
/app/lib/pharos/kube.rb:66:in `create_resource'
/app/lib/pharos/kube.rb:164:in `rescue in apply_resource'
/app/lib/pharos/kube.rb:161:in `apply_resource'
/app/lib/pharos/kube.rb:121:in `block in apply_stack'
/app/lib/pharos/kube.rb:115:in `each'
/app/lib/pharos/kube.rb:115:in `apply_stack'
/app/lib/pharos/addon.rb:97:in `apply_stack'
/app/lib/pharos/addons/ingress_nginx.rb:38:in `install'
/app/lib/pharos/addon_manager.rb:35:in `block in apply'
/app/lib/pharos/addon_manager.rb:49:in `block in with_enabled_addons'
/app/lib/pharos/addon_manager.rb:46:in `each'
/app/lib/pharos/addon_manager.rb:46:in `with_enabled_addons'
/app/lib/pharos/addon_manager.rb:31:in `apply'
/app/lib/pharos/up_command.rb:162:in `handle_addons'
/app/lib/pharos/up_command.rb:46:in `block in configure'
/app/lib/pharos/up_command.rb:43:in `chdir'
/app/lib/pharos/up_command.rb:43:in `configure'
/app/lib/pharos/up_command.rb:27:in `execute'
/usr/local/bundle/gems/clamp-1.2.1/lib/clamp/command.rb:63:in `run'
/usr/local/bundle/gems/clamp-1.2.1/lib/clamp/subcommand/execution.rb:11:in `execute'
/usr/local/bundle/gems/clamp-1.2.1/lib/clamp/command.rb:63:in `run'
/usr/local/bundle/gems/clamp-1.2.1/lib/clamp/command.rb:132:in `run'
/app/lib/pharos/root_command.rb:14:in `run'
bin/pharos-cluster:12:in `<main>'
```

### PR DESCRIPTION
**Pull request by kke**
Fixes #26 
Fixes #199
Replaces #47 
Includes #52 
Replaces / includes #200

- Only process ERB when filename ends with .erb or forced (`force_erb: true`)
- Replace `Pharos::Erb` with `Pharos::YamlFile`
- Use a custom `Pharos::YamlFile::Namespace` to replace the `OpenStruct` for ERB variables, makes ERB fail for unknown variables instead of passing silently with empty strings
- Looks for cluster.yml.erb in addition to cluster.yml if config file not specified
- Adds file paths to YAML and ERB parse error messages
- Unifies the way YAML.safe_load is used (no white/blacklisted classes, support anchors, pass filename)

### CODE REVIEW

**Comment by kke** [comment]
Quite a horrible rebase with the renames.

File: lib/pharos/kube.rb [hunk 1]
```diff
@@ -212,10 +228,9 @@ def self.delete_resource(host, resource)
     # @param path [String]
     # @return [Kubeclient::Resource]
     def self.parse_resource_file(path, vars = {})
-      yaml = File.read(File.realpath(File.join(__dir__, 'resources', path)))
-      parsed_yaml = Pharos::Erb.new(yaml).render(vars)
-
-      Kubeclient::Resource.new(YAML.safe_load(parsed_yaml))
+      Kubeclient::Resource.new(
+        Pharos::YamlFile.new(File.realpath(resource_path(path))).load(vars)
```
Comment (SpComb) [inline lib/pharos/kube.rb]: The arg be something like `filename` instead of `path`, if `resource_files` => `Dir.glob` only returns the base filename.

Better yet, have `resource_files` just return the full paths.

File: lib/pharos/addons/cert_manager.rb [hunk 2]
```diff
@@ -9,7 +9,7 @@ class CertManager < Pharos::Addon
 
       class Issuer < Pharos::Addons::Struct
         attribute :name, Pharos::Types::String
-        attribute :server, Pharos::Types::String
+        attribute :server, Pharos::Types::String.optional
```
Comment (SpComb) [inline lib/pharos/addons/cert_manager.rb]: Unrelated change?

File: lib/pharos/yaml_file.rb [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    # @param input [String,IO] A IO/File object, a path to a file or string content
+    # @param override_filename [String] use string as the filename for parse errors
+    # @param force_erb [Boolean] force erb processing even if filename does not end in .yml or is unknown
+    def initialize(input, override_filename: nil, force_erb: false)
+      @filename = override_filename
+      if input.respond_to?(:read)
+        @content = input.read
+        @filename ||= input.respond_to?(:path) ? input.path : nil
+      elsif input.include?("\n")
+        @content = input
```
Comment (SpComb) [inline lib/pharos/yaml_file.rb]: Too magic, I think.. better to wrap the input in a `StringIO` instead?

File: lib/pharos/yaml_file.rb [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+    def read(variables = {})
+      erb? ? erb_result(variables) : @content
+    end
+
+    def erb_result(variables = {})
+      Namespace.new(@filename, variables).with_binding do |ns_binding|
+        ERB.new(@content, nil, '%<>-').result(ns_binding)
+      end
+    rescue NameSpace::Error
```
Comment (SpComb) [inline lib/pharos/yaml_file.rb]:
```
uninitialized constant Pharos::YamlFile::NameSpace
Did you mean?  Pharos::YamlFile::Namespace
/app/lib/pharos/yaml_file.rb:57:in `rescue in erb_result'
/app/lib/pharos/yaml_file.rb:53:in `erb_result'
/app/lib/pharos/yaml_file.rb:50:in `read'
/app/lib/pharos/yaml_file.rb:32:in `load'
/app/lib/pharos/kube.rb:232:in `parse_resource_file'
/app/lib/pharos/kube.rb:129:in `block in apply_stack'
/app/lib/pharos/kube.rb:128:in `each'
/app/lib/pharos/kube.rb:128:in `apply_stack'
/app/lib/pharos/phases/configure_network.rb:50:in `ensure_resources'
/app/lib/pharos/phases/configure_network.rb:25:in `call'
/app/lib/pharos/up_command.rb:131:in `handle_masters'
/app/lib/pharos/up_command.rb:42:in `block in configure'
/app/lib/pharos/up_command.rb:41:in `chdir'
/app/lib/pharos/up_command.rb:41:in `configure'
/app/lib/pharos/up_command.rb:25:in `execute'
/usr/local/bundle/gems/clamp-1.2.1/lib/clamp/command.rb:63:in `run'
/usr/local/bundle/gems/clamp-1.2.1/lib/clamp/subcommand/execution.rb:11:in `execute'
/usr/local/bundle/gems/clamp-1.2.1/lib/clamp/command.rb:63:in `run'
/usr/local/bundle/gems/clamp-1.2.1/lib/clamp/command.rb:132:in `run'
/app/lib/pharos/root_command.rb:14:in `run'
bin/pharos-cluster:12:in `<main>'
```

File: lib/pharos/yaml_file.rb [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+      @filename.nil? ? '<unknown>' : File.basename(@filename)
+    end
+
+    def read(variables = {})
+      erb? ? erb_result(variables) : @content
+    end
+
+    def erb_result(variables = {})
+      Namespace.new(@filename, variables).with_binding do |ns_binding|
+        ERB.new(@content, nil, '%<>-').result(ns_binding)
```
Comment (SpComb) [inline lib/pharos/yaml_file.rb]: Is this missing the `erb.location = [@filename, nil]` for syntax errors, like in #125?

The `rescue` bits add in the filename, but it would be better to get that immediately in the `(erb):121` error with the correct line numbering.

```
(erb):121: syntax error, unexpected end-of-input, expecting keyword_end
t.force_encoding(__ENCODING__)
                              ^ : (erb):121: syntax error, unexpected end-of-input, expecting keyword_end
t.force_encoding(__ENCODING__)
                              ^ in file /app/lib/pharos/resources/weave/daemon-set.yml.erb
/app/lib/pharos/yaml_file.rb:60:in `rescue in erb_result'
/app/lib/pharos/yaml_file.rb:53:in `erb_result'
/app/lib/pharos/yaml_file.rb:50:in `read'
/app/lib/pharos/yaml_file.rb:32:in `load'
/app/lib/pharos/kube.rb:232:in `parse_resource_file'
/app/lib/pharos/kube.rb:129:in `block in apply_stack'
/app/lib/pharos/kube.rb:128:in `each'
/app/lib/pharos/kube.rb:128:in `apply_stack'
/app/lib/pharos/phases/configure_network.rb:50:in `ensure_resources'
/app/lib/pharos/phases/configure_network.rb:25:in `call'
/app/lib/pharos/up_command.rb:131:in `handle_masters'
/app/lib/pharos/up_command.rb:42:in `block in configure'
/app/lib/pharos/up_command.rb:41:in `chdir'
/app/lib/pharos/up_command.rb:41:in `configure'
/app/lib/pharos/up_command.rb:25:in `execute'
/usr/local/bundle/gems/clamp-1.2.1/lib/clamp/command.rb:63:in `run'
/usr/local/bundle/gems/clamp-1.2.1/lib/clamp/subcommand/execution.rb:11:in `execute'
/usr/local/bundle/gems/clamp-1.2.1/lib/clamp/command.rb:63:in `run'
/usr/local/bundle/gems/clamp-1.2.1/lib/clamp/command.rb:132:in `run'
/app/lib/pharos/root_command.rb:14:in `run'
bin/pharos-cluster:12:in `<main>'
```

File: lib/pharos/yaml_file/namespace.rb [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+      end
+
+      def inspect
+        super.sub('>', "for='#{@path}'>")
+      end
+
+      def with_binding(&block)
+        yield binding
+      rescue NameError => ex
+        raise Error, "#{ex.message} in file #{@path}"
```
Comment (SpComb) [inline lib/pharos/yaml_file/namespace.rb]: Not sure if this exception wrapping is a good idea, this is missing the line number... I think erb itself would be able to handle the `NameError` better if you tell it the `#location=`?

```
undefined local variable or method `asdf' for #<Pharos::YamlFile::Namespace:0x000000000324c568> in file /app/lib/pharos/resources/weave/daemon-set.yml.erb
/app/lib/pharos/yaml_file/namespace.rb:22:in `rescue in with_binding'
/app/lib/pharos/yaml_file/namespace.rb:19:in `with_binding'
/app/lib/pharos/yaml_file.rb:54:in `erb_result'
/app/lib/pharos/yaml_file.rb:50:in `read'
/app/lib/pharos/yaml_file.rb:32:in `load'
/app/lib/pharos/kube.rb:232:in `parse_resource_file'
/app/lib/pharos/kube.rb:129:in `block in apply_stack'
/app/lib/pharos/kube.rb:128:in `each'
/app/lib/pharos/kube.rb:128:in `apply_stack'
/app/lib/pharos/phases/configure_network.rb:50:in `ensure_resources'
/app/lib/pharos/phases/configure_network.rb:25:in `call'
/app/lib/pharos/up_command.rb:131:in `handle_masters'
/app/lib/pharos/up_command.rb:42:in `block in configure'
/app/lib/pharos/up_command.rb:41:in `chdir'
/app/lib/pharos/up_command.rb:41:in `configure'
/app/lib/pharos/up_command.rb:25:in `execute'
/usr/local/bundle/gems/clamp-1.2.1/lib/clamp/command.rb:63:in `run'
/usr/local/bundle/gems/clamp-1.2.1/lib/clamp/subcommand/execution.rb:11:in `execute'
/usr/local/bundle/gems/clamp-1.2.1/lib/clamp/command.rb:63:in `run'
/usr/local/bundle/gems/clamp-1.2.1/lib/clamp/command.rb:132:in `run'
/app/lib/pharos/root_command.rb:14:in `run'
bin/pharos-cluster:12:in `<main>'
```

File: lib/pharos/yaml_file.rb [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+      end
+    rescue NameSpace::Error
+      raise
+    rescue StandardError, ScriptError => ex
+      raise ParseError, "#{ex} : #{ex.message} in file #{@filename}"
```
Comment (SpComb) [inline lib/pharos/yaml_file.rb]: The `#{ex} : #{ex.message}` prints the message twice? Should that be `ex.class`?

```
undefined local variable or method `asdf' for #<Pharos::YamlFile::Namespace:0x0000000003ba8490> : undefined local variable or method `asdf' for #<Pharos::YamlFile::Namespace:0x0000000003ba8490> in file /app/lib/pharos/resources/weave/daemon-set.yml.erb
```

File: lib/pharos/yaml_file/namespace.rb (hunk 6 above)
Comment (SpComb) [reply lib/pharos/yaml_file/namespace.rb]: For comparison, removing all the rescue-and-raise bits and modifying the `erb_result` to set the `#location=` instead, the resulting NameError is more useful as it also contains the erb template line number:

```
undefined local variable or method `asdf' for #<Pharos::YamlFile::Namespace:0x0000000001ed8c48>
/app/lib/pharos/resources/weave/daemon-set.yml.erb:30:in `with_binding'
/usr/local/lib/ruby/2.4.0/erb.rb:896:in `eval'
/usr/local/lib/ruby/2.4.0/erb.rb:896:in `result'
/app/lib/pharos/yaml_file.rb:57:in `block in erb_result'
/app/lib/pharos/yaml_file/namespace.rb:20:in `with_binding'
/app/lib/pharos/yaml_file.rb:54:in `erb_result'
/app/lib/pharos/yaml_file.rb:50:in `read'
/app/lib/pharos/yaml_file.rb:32:in `load'
/app/lib/pharos/kube.rb:232:in `parse_resource_file'
/app/lib/pharos/kube.rb:129:in `block in apply_stack'
/app/lib/pharos/kube.rb:128:in `each'
/app/lib/pharos/kube.rb:128:in `apply_stack'
/app/lib/pharos/phases/configure_network.rb:50:in `ensure_resources'
/app/lib/pharos/phases/configure_network.rb:25:in `call'
/app/lib/pharos/up_command.rb:131:in `handle_masters'
/app/lib/pharos/up_command.rb:42:in `block in configure'
/app/lib/pharos/up_command.rb:41:in `chdir'
/app/lib/pharos/up_command.rb:41:in `configure'
/app/lib/pharos/up_command.rb:25:in `execute'
/usr/local/bundle/gems/clamp-1.2.1/lib/clamp/command.rb:63:in `run'
/usr/local/bundle/gems/clamp-1.2.1/lib/clamp/subcommand/execution.rb:11:in `execute'
/usr/local/bundle/gems/clamp-1.2.1/lib/clamp/command.rb:63:in `run'
/usr/local/bundle/gems/clamp-1.2.1/lib/clamp/command.rb:132:in `run'
/app/lib/pharos/root_command.rb:14:in `run'
bin/pharos-cluster:12:in `<main>'
```

File: lib/pharos/yaml_file.rb (hunk 5 above)
Comment (SpComb) [reply lib/pharos/yaml_file.rb]:
```patch
diff --git a/lib/pharos/yaml_file.rb b/lib/pharos/yaml_file.rb
index 2bdb91c..dfceba9 100644
--- a/lib/pharos/yaml_file.rb
+++ b/lib/pharos/yaml_file.rb
@@ -52,12 +52,10 @@ module Pharos
 
     def erb_result(variables = {})
       Namespace.new(@filename, variables).with_binding do |ns_binding|
-        ERB.new(@content, nil, '%<>-').result(ns_binding)
+        erb = ERB.new(@content, nil, '%<>-')
+        erb.location = [@filename, nil]
+        erb.result(ns_binding)
       end
-    rescue NameSpace::Error
-      raise
-    rescue StandardError, ScriptError => ex
-      raise ParseError, "#{ex} : #{ex.message} in file #{@filename}"
     end
 
     private
diff --git a/lib/pharos/yaml_file/namespace.rb b/lib/pharos/yaml_file/namespace.rb
index 16093a8..9866fb8 100644
--- a/lib/pharos/yaml_file/namespace.rb
+++ b/lib/pharos/yaml_file/namespace.rb
@@ -18,8 +18,6 @@ module Pharos
 
       def with_binding(&block)
         yield binding
-      rescue NameError => ex
-        raise Error, "#{ex.message} in file #{@path}"
       end
     end
   end
```

File: lib/pharos/addons/cert_manager.rb (hunk 2 above)
Comment (kke) [reply lib/pharos/addons/cert_manager.rb]: Yes / No. It's optional in the schema below, so it's a bug on it's own, but with the namespace thing it makes it explode unless it's filled.

File: lib/pharos/kube.rb (hunk 1 above)
Comment (kke) [reply lib/pharos/kube.rb]: Hmm?

```
irb(main):005:0> Dir.glob('/tmp/*')
=> ["/tmp/12b91f1745676f37ea3b81e48869fd3c-{87A94AB0-E370-4cde-98D3-ACC110C5967D}", "/tmp/com.apple.launchd.SjQyHUEOBP", "/tmp/com.apple.launchd.zV0kvlYB3q", "/tmp/jorma"]
```

Comment (SpComb) [reply lib/pharos/kube.rb]: Then why the `resource_path(path)`?

I was looking at the examples in the docs too literally...

File: lib/pharos/addons/cert_manager.rb (hunk 2 above)
Comment (SpComb) [reply lib/pharos/addons/cert_manager.rb]: Right, these need stricter validation now... erb templates defaulted typo'd variables to `nil`.

Comment (SpComb) [reply lib/pharos/addons/cert_manager.rb]: BTW strong :+1: for this change, because the current behavior means that typos in erb templates or changes to `apply_stack` options can silently break... something like the following would end up deploying weave without encryption, which would suck hard:

```patch
diff --git a/lib/pharos/resources/weave/daemon-set.yml b/lib/pharos/resources/weave/daemon-set.yml
index 50f6c52..8a18032 100644
--- a/lib/pharos/resources/weave/daemon-set.yml
+++ b/lib/pharos/resources/weave/daemon-set.yml
@@ -27,7 +27,7 @@ spec:
               value: "0"
             - name: CHECKPOINT_DISABLE
               value: "1"
-            <% if !trusted_subnets.empty? %>
+            <% if !trusted_sbunets.empty? %>
             - name: EXTRA_ARGS
               value: "--trusted-subnets <%= trusted_subnets.join(',') %>"
             <% end %>
```

File: lib/pharos/kube.rb (hunk 1 above)
Comment (kke) [reply lib/pharos/kube.rb]: I think maybe mostly because of ./spec/pharos/kube_spec.rb:4

```ruby
      it 'returns resource' do
        resource = described_class.parse_resource_file('ingress-nginx/03-role.yml')
        expect(resource.metadata.name).to eq('nginx-ingress-role')
      end
```

which called it using a relative path.

File: lib/pharos/kube.rb [hunk 8]
```diff
@@ -126,7 +126,7 @@ def self.apply_stack(host, stack, vars = {})
       checksum = SecureRandom.hex(16)
       resources = []
       resource_files(stack).each do |file|
-        resource = parse_resource_file("#{stack}/#{File.basename(file)}", vars)
+        resource = parse_resource_file(file, vars)
```
Comment (SpComb) [inline lib/pharos/kube.rb]: Much better

**Comment by kke** [comment]
Weird. Pathname.new('/tmp').glob('*') says `undefined method glob for #<Pathname..>` on 2.4.2 but not on 2.5.1. from the docs it looks like it was already there already in 1.8.7.

File: lib/pharos/kube.rb [hunk 9]
```diff
@@ -115,7 +115,7 @@ def self.resource_path(*joinables)
     # @param stack [String]
     # @return [Array<Pathname>]
     def self.resource_files(stack)
-      Pathname.glob(resource_path(stack, '*.{yml,yml.erb}'))
+      Pathname.glob(resource_path(stack, '*.{yml,yml.erb}')).sort_by(&:to_s)
```
Comment (SpComb) [inline lib/pharos/kube.rb]: This fixes #199

Comment (SpComb) [reply lib/pharos/kube.rb]: #200 for a simplified version of this.
